### PR TITLE
3877, 3892, 3814 and 3876

### PR DESCRIPTION
**Pull request by julianhyde**

### CODE REVIEW

File: core/src/main/java/org/apache/calcite/rel/rel2sql/SqlImplementor.java [hunk 1]
```diff
@@ -1439,6 +1439,14 @@ private boolean needNewSubQuery(RelNode rel, Clause[] clauses) {
         return true;
       }
 
+      if (rel instanceof Project
```
Comment (DonnyZone) [inline core/src/main/java/org/apache/calcite/rel/rel2sql/SqlImplementor.java:1442]: Do we also need to prevent merging projects in RelBuilder when converting `SqlNode` to `RelNode`?

```
LogicalProject(EXPR$0=[ROW_NUMBER() OVER (ORDER BY ROW_NUMBER() OVER (ORDER BY $1 ROWS BETWEEN UNBOUNDED PRECEDING AND CURRENT ROW) ROWS BETWEEN UNBOUNDED PRECEDING AND CURRENT ROW)])
  JdbcTableScan(table=[[foodmart, product]])
```

File: core/src/test/java/org/apache/calcite/rel/rel2sql/RelToSqlConverterTest.java [hunk 2]
```diff
@@ -2749,6 +2749,28 @@ private void checkLiteral2(String expression, String expected) {
...
   }
 
+  /** Test case for
+   * <a href="https://issues.apache.org/jira/browse/CALCITE-3876">[CALCITE-3876]
+   * RelToSqlConverter should not combine Projects when top Project contains
+   * window function referencing window function from bottom Project</a>. */
+  @Test public void testWindowOnWindowDoesNotCombineProjects() {
+    final String query = "SELECT ROW_NUMBER() OVER (ORDER BY rn)\n"
+        + "FROM (SELECT *,\n"
+        + "  ROW_NUMBER() OVER (ORDER BY \"product_id\") as rn\n"
```
Comment (danny0405) [inline core/src/test/java/org/apache/calcite/rel/rel2sql/RelToSqlConverterTest.java:2759]: What about we do not have any options for `OVER `, there would then has no reference for underlying inputs.

File: core/src/main/java/org/apache/calcite/sql2rel/SqlToRelConverter.java [hunk 3]
```diff
@@ -1943,24 +1944,43 @@ private RexNode convertOver(Blackboard bb, SqlNode node) {
...
+      if (aggCall.getKind() == SqlKind.ROW_NUMBER) {
+        // ROW_NUMBER() expects specific kind of framing.
+        rows = true;
+      }
+    } else if (orderList.size() == 0) {
+      // Without ORDER BY, there must be no bracketing.
+      sqlLowerBound = SqlWindow.createUnboundedPreceding(SqlParserPos.ZERO);
+      sqlUpperBound = SqlWindow.createUnboundedFollowing(SqlParserPos.ZERO);
+    } else if (sqlLowerBound == null && sqlUpperBound == null) {
+      sqlLowerBound = SqlWindow.createUnboundedPreceding(SqlParserPos.ZERO);
```
Comment (danny0405) [inline core/src/main/java/org/apache/calcite/sql2rel/SqlToRelConverter.java:1968]: This block seems not necessary.

File: core/src/main/java/org/apache/calcite/sql2rel/SqlToRelConverter.java [hunk 4]
```diff
@@ -1943,24 +1944,43 @@ private RexNode convertOver(Blackboard bb, SqlNode node) {
...
+      // everything up to the current row.
+      sqlLowerBound = SqlWindow.createUnboundedPreceding(SqlParserPos.ZERO);
+      sqlUpperBound = SqlWindow.createCurrentRow(SqlParserPos.ZERO);
+      if (aggCall.getKind() == SqlKind.ROW_NUMBER) {
+        // ROW_NUMBER() expects specific kind of framing.
+        rows = true;
+      }
+    } else if (orderList.size() == 0) {
+      // Without ORDER BY, there must be no bracketing.
+      sqlLowerBound = SqlWindow.createUnboundedPreceding(SqlParserPos.ZERO);
```
Comment (danny0405) [inline core/src/main/java/org/apache/calcite/sql2rel/SqlToRelConverter.java:1965]: Is this true ? From the MS-SQL syntax, the `ORDER BY` has no relationship with `<ROW or RANGE clause>`, they can both exist separately. [1] 

https://docs.microsoft.com/en-us/sql/t-sql/queries/select-over-clause-transact-sql?view=sql-server-ver15

File: core/src/main/java/org/apache/calcite/schema/Schemas.java [hunk 5]
```diff
@@ -224,15 +224,15 @@ public static DataContext createDataContext(
...
   }
 
   /** Returns an {@link org.apache.calcite.linq4j.Enumerable} over the rows of
    * a given table, not applying any filters and projecting all columns,
    * representing each row as an object array. */
   public static Enumerable<Object[]> enumerable(
       final ProjectableFilterableTable table, final DataContext root) {
-    return table.scan(root, ImmutableList.of(),
+    return table.scan(root, new ArrayList<>(),
         identity(table.getRowType(root.getTypeFactory()).getFieldCount()));
```
Comment (danny0405) [inline core/src/main/java/org/apache/calcite/schema/Schemas.java:236]: Why this change ?

**Comment by vlsi** [comment]
@julianhyde , please split "Make junit test classes and methods non-public where p…" into its own pull request.

"make junit test non-public" adds very little value, however, it makes it **extremely** complicated to review and comment on other changes.

File: core/src/main/java/org/apache/calcite/schema/Schemas.java (hunk 5 above)
Comment (vlsi) [reply core/src/main/java/org/apache/calcite/schema/Schemas.java:236]: I have the same question. If `ImmutableList.of()` is bad, then we should forbid its use, and move to `Collections.emptyList()`.

File: site/_docs/history.md [hunk 6]
```diff
@@ -31,6 +31,11 @@ Downloads are available on the
 ## <a href="https://github.com/apache/calcite/releases/tag/calcite-1.23.0">1.23.0</a> / under development
 {: #v1-23-0}
 
+Compatibility: This release is tested on Linux, macOS, Microsoft Windows;
+using Oracle JDK 8, 9, 10, 11, 12, 13, 14 and OpenJDK 8, 9, 10, 11, 12, 13, 14;
```
Comment (vlsi) [inline site/_docs/history.md:35]: Please remove 9, 10, 11, 12, 13 from the list. We test only LTS (8, 11) and EA (14) JDK versions.
I'm not sure we test with Oracle JDK.

File: site/_docs/history.md [hunk 7]
```diff
@@ -31,6 +31,11 @@ Downloads are available on the
...
+Guava versions 19.0 to 28.2-jre; Apache Flink 1.10.0;
```
Comment (vlsi) [inline site/_docs/history.md:36]: > 19.0 to 28.2-jre;

Are you sure all the versions in-between are really tested? Where's the CI?

File: core/src/main/java/org/apache/calcite/schema/Schemas.java [hunk 8]
```diff
@@ -224,15 +224,15 @@ public static DataContext createDataContext(
    * array. */
   public static Enumerable<Object[]> enumerable(final FilterableTable table,
       final DataContext root) {
-    return table.scan(root, ImmutableList.of());
+    return table.scan(root, new ArrayList<>());
```
Comment (vlsi) [inline core/src/main/java/org/apache/calcite/schema/Schemas.java:227]:
```suggestion
    return table.scan(root, Collections.emptyList());
```

File: babel/src/test/resources/sql/redshift.iq [hunk 9]
```diff
@@ -603,11 +603,11 @@ EMPNO, EXPR$1
 select dense_rank() over () from emp where deptno = 30;
 EXPR$0
 6
-6
-6
-6
-6
-6
+1
```
Comment (vlsi) [inline babel/src/test/resources/sql/redshift.iq:606]: What does dense_rank without ORDER BY mean?
I guess it should be an error.

File: core/src/main/java/org/apache/calcite/rel/core/Window.java [hunk 10]
```diff
@@ -246,26 +244,38 @@ public String toString() {
...
+          && lowerBound.isUnbounded()
+          && lowerBound.isPreceding()
+          && upperBound.isUnbounded()
+          && upperBound.isFollowing()) {
+        // skip bracket if no ORDER BY, and if bracket is the default
+      } else if (!orderKeys.getFieldCollations().isEmpty()
+          && lowerBound.isUnbounded()
+          && lowerBound.isPreceding()
+          && upperBound.isCurrentRow()) {
+        // skip bracket if there is ORDER BY, and if bracket is the default
```
Comment (vlsi) [inline core/src/main/java/org/apache/calcite/rel/core/Window.java:267]: This is invalid because it does not verify if `rows` vs `range` is in its default value.

For instance,
`count(x) over (order by x RANGE between unbounded preceding and current row)` returns different values depending on RANGE vs ROW

File: core/src/main/java/org/apache/calcite/rel/rel2sql/SqlImplementor.java (hunk 1 above)
Comment (julianhyde) [reply core/src/main/java/org/apache/calcite/rel/rel2sql/SqlImplementor.java:1442]: Yes, but that's covered in https://issues.apache.org/jira/browse/CALCITE-3079.

File: core/src/main/java/org/apache/calcite/sql2rel/SqlToRelConverter.java (hunk 4 above)
Comment (julianhyde) [reply core/src/main/java/org/apache/calcite/sql2rel/SqlToRelConverter.java:1965]: How can you say '5 ROWS PRECEDING' if you don't provide an ORDER BY clause to say by which criteria they are preceding?

If rows/range are present and there is no ORDER BY the validator doesn't throw (maybe it should) but we treat all rows as equal.

File: core/src/main/java/org/apache/calcite/schema/Schemas.java (hunk 5 above)
Comment (julianhyde) [reply core/src/main/java/org/apache/calcite/schema/Schemas.java:236]: See the commit comment. That parameter to `scan` is mutable, always has been. The implementation calls `List.removeIf` on it. As of guava-28.2-jre, `removeIf` throws even if the list is empty. So, these calls were incorrect to pass an immutable list.

File: site/_docs/history.md (hunk 6 above)
Comment (julianhyde) [reply site/_docs/history.md:35]: I test on all of those JDKs.

File: site/_docs/history.md (hunk 7 above)
Comment (julianhyde) [reply site/_docs/history.md:36]: Not sure where 'Apache Flink' came from. There used to be 'Apache Druid', because we have a Druid adapter, and we tested against a particular version of Druid. We don't depend on any Druid maven artifacts - because we speak to Druid using REST - so the version is not apparent by simply looking at our maven dependencies.

Flink doesn't belong on this list because we don't test against Flink. Flink tests against us.

@danny0405 Can you comment?

File: core/src/main/java/org/apache/calcite/schema/Schemas.java (hunk 8 above)
Comment (julianhyde) [reply core/src/main/java/org/apache/calcite/schema/Schemas.java:227]: No. We need a mutable list. Try running with `-Pguava.version=28.2-jre` and you will see.

File: babel/src/test/resources/sql/redshift.iq (hunk 9 above)
Comment (julianhyde) [reply babel/src/test/resources/sql/redshift.iq:606]: It is an error in Calcite (I checked), but not in Redshift.

File: core/src/main/java/org/apache/calcite/schema/Schemas.java (hunk 8 above)
Comment (vlsi) [reply core/src/main/java/org/apache/calcite/schema/Schemas.java:227]: Why do you think mutable list is required here?
It was working with `ImmutableList.of()`, so it means it should be perfectly fine to pass `Collections.emptyList()`.

File: core/src/main/java/org/apache/calcite/schema/Schemas.java (hunk 5 above)
Comment (vlsi) [reply core/src/main/java/org/apache/calcite/schema/Schemas.java:236]: @julianhyde , `Collection.emptyList()` supports `removeIf` just fine.

File: babel/src/test/resources/sql/redshift.iq (hunk 9 above)
Comment (vlsi) [reply babel/src/test/resources/sql/redshift.iq:606]: Is Redshift different from PostgreSQL?

```sql
select dense_rank() over() from (select * from generate_series(1,5)) y
```

returns 1,1,1,1,1;

It is really strange to see `1,2,3,4,5` from `dense_rank()`

File: core/src/main/java/org/apache/calcite/schema/Schemas.java (hunk 8 above)
Comment (julianhyde) [reply core/src/main/java/org/apache/calcite/schema/Schemas.java:227]: The javadoc says "the list of filters is mutable".

File: babel/src/test/resources/sql/redshift.iq (hunk 9 above)
Comment (julianhyde) [reply babel/src/test/resources/sql/redshift.iq:606]: I don't know. Since this is babel, the main point of the test is parsing & validation, not execution. I don't much care if the results are wrong.

File: core/src/main/java/org/apache/calcite/schema/Schemas.java (hunk 5 above)
Comment (julianhyde) [reply core/src/main/java/org/apache/calcite/schema/Schemas.java:236]: But it's not mutable. The contract calls to a mutable list.

File: core/src/test/java/org/apache/calcite/rel/rel2sql/RelToSqlConverterTest.java (hunk 2 above)
Comment (julianhyde) [reply core/src/test/java/org/apache/calcite/rel/rel2sql/RelToSqlConverterTest.java:2759]: I think `ROW_NUMBER` must have `ORDER BY` in its window, and if so, this concern is moot.

File: core/src/main/java/org/apache/calcite/rel/core/Window.java (hunk 10 above)
Comment (julianhyde) [reply core/src/main/java/org/apache/calcite/rel/core/Window.java]: Good point. I've fixed in 75ee3ec79.

File: core/src/main/java/org/apache/calcite/sql2rel/SqlToRelConverter.java (hunk 3 above)
Comment (julianhyde) [reply core/src/main/java/org/apache/calcite/sql2rel/SqlToRelConverter.java:1968]: It is necessary - without it, there are failures in SqlToRelConverterTest.

File: site/_docs/history.md (hunk 7 above)
Comment (danny0405) [reply site/_docs/history.md:36]: I test Flink manually when upgrade Calcite for Flink and prepare for Calcite release. And i put it in the release note.

File: core/src/main/java/org/apache/calcite/sql2rel/SqlToRelConverter.java (hunk 3 above)
Comment (danny0405) [reply core/src/main/java/org/apache/calcite/sql2rel/SqlToRelConverter.java:1968]: Yeah, you are right, it is not equivalent with the following block.

File: core/src/main/java/org/apache/calcite/sql2rel/SqlToRelConverter.java (hunk 4 above)
Comment (danny0405) [reply core/src/main/java/org/apache/calcite/sql2rel/SqlToRelConverter.java:1965]: You are right, i checked these for MS-SQL 2017
```sql
create table t(
  a int,
  b int
);

select max(a) over( rows 2 PRECEDING ) from t; -- syntax error throws
select max(a) over( order by b rows 2 PRECEDING ) from t; -- this works
```

Comment (danny0405) [reply core/src/main/java/org/apache/calcite/sql2rel/SqlToRelConverter.java:1965]: Show we throw for non-order by query ?

File: core/src/main/java/org/apache/calcite/schema/Schemas.java (hunk 8 above)
Comment (vlsi) [reply core/src/main/java/org/apache/calcite/schema/Schemas.java:227]: javadoc does not say the implementation can **add** new filters to the list. So `Collections.emptyList()` is fine.

File: core/src/main/java/org/apache/calcite/schema/Schemas.java (hunk 5 above)
Comment (vlsi) [reply core/src/main/java/org/apache/calcite/schema/Schemas.java:236]: If you think the contract is to treat the list as fully mutable, then, adjust the Javadoc, and adjust the implementation so it handles the mutations of the list.

In other words, for now you pass an ArrayList and expect it to be mutated. However, you seem to ignore the mutated value of the ArrayList, so the new code might result in invalid data being returned.

File: site/_docs/history.md (hunk 7 above)
Comment (vlsi) [reply site/_docs/history.md:36]: Calcite does not rely on Flink, so I guess we should remove Flink from the list here because it might confuse the reader.

JDK releases do not put words like "this JDK release was tested with Calcite 1.22"

File: core/src/main/java/org/apache/calcite/sql2rel/SqlToRelConverter.java (hunk 4 above)
Comment (julianhyde) [reply core/src/main/java/org/apache/calcite/sql2rel/SqlToRelConverter.java:1965]: I have logged https://issues.apache.org/jira/browse/CALCITE-3897. We should decide whether we want to fix it.
